### PR TITLE
Add outer indices to threadwise write/read ops

### PR DESCRIPTION
**Pull request by manupak**
This commit adds support for threadwise write/read ops to accept a additional coordinates that is already determined by the client op that creates them.

E.g. in a tiled gemm implementation the loaded values will be viewed as (kOuter, bid, tid, iter). However, kOuter will be iterated by the outerloop in a thread. We can use threadwise_write_all within the loop where kOuter, bid, and iter as an extra indices passed onto threadwise_write_all.

closes : https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/1009

### CODE REVIEW

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 1]
```diff
@@ -858,14 +872,23 @@ LogicalResult ThreadwiseWriteAllRewritePattern::matchAndRewrite(
...
 
-  SmallVector<Value, 3> writeStartCoords{tid, zero};
-  SmallVector<int64_t, 3> bounds{1, iterLen};
-  SmallVector<int64_t, 3> strides{1, vectorLen};
+  SmallVector<Value, 3> writeStartCoords;
+  SmallVector<int64_t, 3> bounds;
+  SmallVector<int64_t, 3> strides;
+  ValueRange extraIndices = op.getExtraIndices();
+  for (Value extraIdx : extraIndices) {
+    writeStartCoords.push_back(extraIdx);
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: Nit: There are constructors for both copying in a range and producing a vector of N copies of something.

**Review by krzysz00 — COMMENTED** [review]
As a general thought, can we actually make the block ID one of these extra indices instead of making it implicit? This'll also simplify @giuseros's layout refactor, and make threadwise_write_all less special-case-y.

**Comment by manupak** [comment]
Hmmm.... why not 'tid' too ? The lowering doesn't do anything special as compared to 'bid'.
I mean if the client is creating/passing the views, they might give the 'bid', 'tid' too and let the op handle 'iter'

**Comment by krzysz00** [comment]
.. yeah, passing in `tid` isn't a bad call.

**Comment by manupak** [comment]
ok, I've generalized it now. PTAL.

**Comment by manupak** [comment]
A gentle ping for reviews!

**Review by krzysz00 — APPROVED** [review]
Terribly sorry for forgetting to review, great cleanup, approved!

**Comment by manupak** [comment]
Thanks @krzysz00!

**Pull request merged**